### PR TITLE
WLTS-17 Functional Tests for the new PUT /dictionary endpoint

### PR DESCRIPTION
**Pull request by CameronGuthrie1**
### JIRA link (if applicable) ###
[Functional Tests for the new PUT /dictionary endpoint](https://tools.hmcts.net/jira/browse/WLTS-17)

### Change description ###

The Functional tests for WLTS-4 and WLTS-20

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    "parentContext.testData.request.body.translations"
+                );
+            //This is required to fill in the blank Welsh phrase in
+            //order to match the translation endpoint response
+            return translations.entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                    Map.Entry::getKey,
+                    e -> {
+                        System.out.println("VALUE" + e.getValue());
```
Comment (sullivangareth) [inline src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java:33]: Remove system outs

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+    @Override
+    public Boolean matches(CustomValueKey key) {
+        return CustomValueKey.UNIQUE_LOAD_TRANSLATION.equals(key);
+    }
+
+    @Override
+    public Object calculate(BackEndFunctionalTestScenarioContext scenarioContext, Object key) {
+        String englishPhrase = StringGenerator.generate();
+        return Map.of(englishPhrase, "");
```
Comment (sullivangareth) [inline src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java:17]: `return Map.of(StringGenerator.generate(), "");`

File: src/functionalTest/resources/features/F-004 - Put Dictionary /F-004_Verify.td.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "_extends_": "Common_Request",
+    "headers": {
+      "Accept": "application/json"
+    },
+    "body": {
+      "phrases": "${[scenarioContext][customValues][getUniqueTranslations]}"
+    }
+  },
+  "expectedResponse": {
+    "_extends_": "Common_Response",
```
Comment (sullivangareth) [inline src/functionalTest/resources/features/F-004 - Put Dictionary /F-004_Verify.td.json:24]: ` "_extends_": "Common_200_Response"` ?

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      "translations": {
+        "EnglishPhrase1": "WelshPhrase1",
+        "EnglishPhrase2": "WelshPhrase2",
+        "EnglishPhrase3": "WelshPhrase3"
+      }
+    }
+  },
+
+  "expectedResponse": {
+    "responseCode": 403,
```
Comment (sullivangareth) [inline src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json:28]: `"_extends_": "Common_403_Response"`

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.8_Existing_Data.td.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  "_guid_": "S-004.8_Existing_Data",
+  "_extends_": "Put_Dictionary_Base",
```
Comment (sullivangareth) [inline src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.8_Existing_Data.td.json:3]: If you change to `"_extends_": "Put_Dictionary_Base_Definition_Store"`, you can then delete line 14 ?

File: src/functionalTest/resources/features/common/ts_endpoints/Put_Dictionary_Base_Definition_Store.td.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "_guid_": "Put_Dictionary_Base_Definition_Store",
```
Comment (sullivangareth) [inline src/functionalTest/resources/features/common/ts_endpoints/Put_Dictionary_Base_Definition_Store.td.json:2]: Nothing is currently using this, it look like

S-004.8_Existing_Data.td.json
S-004.8.td.json
S-004.9.td.json

should all be extending from it (and therefore no need for the `"s2sClientId": "ccd_definition",` line in the above files)

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java (hunk 2 above)
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java:17]: done

File: src/functionalTest/resources/features/F-004 - Put Dictionary /F-004_Verify.td.json (hunk 3 above)
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /F-004_Verify.td.json:24]: done

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json (hunk 4 above)
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json:28]: For some reason this does not work

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.8_Existing_Data.td.json (hunk 5 above)
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.8_Existing_Data.td.json:3]: I actually meant to delete that file so all I had to do was add one line to this one

File: src/functionalTest/resources/features/common/ts_endpoints/Put_Dictionary_Base_Definition_Store.td.json (hunk 6 above)
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/common/ts_endpoints/Put_Dictionary_Base_Definition_Store.td.json:2]: I forgot to remove this file so it is now remove

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java (hunk 1 above)
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java]: done

File: src/functionalTest/resources/features/F-004 - Put Dictionary /F-004.feature [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    And   the response has all other details as expected.
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+  @S-004.8 #WLTS-20 AC1
+  Scenario: Do nothing and return success - when service is definition store
+
+    Given a user [without manage-translation or load-translation IDAM role and uses ccd-definition],
+    And   a successful call [to PUT translation phrases into the dictionary] as in [S-004.8_Existing_Data],
+    When  a request is prepared with appropriate values
```
Comment (mattnayler) [inline src/functionalTest/resources/features/F-004 - Put Dictionary /F-004.feature:106]: Can we add an extra spec from the AC to explanin why it is doing nothing: e.g. 
```
AND the request [contain an entry for an english_phrase that already exists] 
```

**Review by mattnayler — COMMENTED** [review]
All tests good, jsut a few minor comments to tweak wording.

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json (hunk 4 above)
Comment (sullivangareth) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json:28]: Try...
 
```
"expectedResponse": {
   "_extends_": "Common_403_Response",
    "body": null
  }
```

Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.4.td.json:28]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/CustomValueKey.java [hunk 8]
```diff
@@ -4,7 +4,11 @@
 
 public enum CustomValueKey {
     UNIQUE_STRING("UniqueString"),
+    UNIQUE_TRANSLATION("UniqueTranslation"),
+    UNIQUE_LOAD_TRANSLATION("UniqueLoadTranslation"),
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/CustomValueKey.java:8]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/CustomValueKey.java [hunk 9]
```diff
@@ -4,7 +4,11 @@
...
     CONTAINS_DICTIONARY_TRANSLATIONS("containsDictionaryTranslations"),
+    GET_UNIQUE_TRANSLATIONS("getUniqueTranslations"),
+    GET_EXPECTED_TRANSLATION_RESPONSE("getExpectedTranslationResponse"),
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/CustomValueKey.java:11]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                    Map.Entry::getKey,
+                    e -> {
+                        if (StringUtils.isBlank(e.getValue())) {
+                            return e.getKey();
+                        }
+                        return e.getValue();
+                    }
+                ));
+        } catch (Exception e) {
+            throw new FunctionalTestException("Problem checking acceptable response payload: ", e);
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java:40]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+            @SuppressWarnings("unchecked") final Map<String, String> translations =
+                (Map<String, String>) ReflectionUtils.deepGetFieldInObject(
+                scenarioContext,
+                "parentContext.testData.request.body.translations"
+            );
+
+            return translations.keySet();
+        } catch (Exception e) {
+            throw new FunctionalTestException("Problem checking acceptable response payload: ", e);
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java:27]: done

File: src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.2.td.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    "invokingUser": {
+      "_extends_": "ManageTranslationsUser"
+    }
+  },
+
+  "request": {
+    "_extends_": "Common_Request",
+    "body": {
+      "translations": {
+        "EnglishPhrase2": "WelshPhrase2"
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /S-004.2.td.json:20]: done

File: src/functionalTest/resources/features/F-004 - Put Dictionary /F-004.feature (hunk 7 above)
Comment (CameronGuthrie1) [reply src/functionalTest/resources/features/F-004 - Put Dictionary /F-004.feature:106]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetExpectedTranslationResponseEvaluator.java]: Sorry, can we use the evalutor name in the error: 
`"Problem generating custom value for GetVerifyTranslationResponseFromParentContextEvaluator: "`

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java]: Sorry, can we use the evalutor name in the error: 
`"Problem generating custom value for getVerifyPhrasesFromParentContext: "`

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+public class UniqueLoadTranslationEvaluator implements CustomValueEvaluator {
+
+    @Override
+    public Boolean matches(CustomValueKey key) {
+        return CustomValueKey.UNIQUE_LOAD_TRANSLATION.equals(key);
+    }
+
+    @Override
+    public Object calculate(BackEndFunctionalTestScenarioContext scenarioContext, Object key) {
+        return Map.of(StringGenerator.generate(), "");
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/UniqueLoadTranslationEvaluator.java]: Closed as null causing errors.

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/GetUniqueTranslationsEvaluator.java]: done

**Review by mattnayler — APPROVED** [review]
LGTM 👍

**Review by mattnayler — APPROVED** [review]
Re-reviewed after rename of FTA folder

**Pull request merged**